### ISSUE TITLE
Update codebase to use ES6 features supported by Node 4.0

### ISSUE THREAD

**Issue by vitorbal** [eslint/eslint#6407]
Continuing the discussion started in #6356, it would be awesome to upgrade our codebase to benefit from ES6 features that are available after dropping Node < 4 support.

Below are some of the ideas already mentioned:

@mikesherov:

> If we're going to start using es6, we should consider using https://github.com/mohebifar/lebab if we're going to upconvert.

@ilyavolodin @platinumazure:

> For dropping Node < 4, should we also convert the code to use some of the ES6 features at the same time, as well as modify our .eslintrc file for this repository? Or should we leave it as is and change it as we go?

@mysticatea:

> I'd like to use `eslint-plugin-node` :innocent:
> 
> - [node/no-deprecated-api](https://github.com/mysticatea/eslint-plugin-node/blob/master/docs/rules/no-deprecated-api.md) - Disallow deprecated API.
> - [node/no-missing-require](https://github.com/mysticatea/eslint-plugin-node/blob/master/docs/rules/no-missing-require.md) - Disallow `require()`s for files that don't exist.
> - [node/no-unpublished-require](https://github.com/mysticatea/eslint-plugin-node/blob/master/docs/rules/no-unpublished-require.md) - Disallow `require()`s for files that are not published.
> - [node/no-unsupported-features](https://github.com/mysticatea/eslint-plugin-node/blob/master/docs/rules/no-unsupported-features.md) - Disallow unsupported ECMAScript features on the specified version.
> 
> Especially, `node/no-unsupported-features` would help us.

**Comment by nzakas** [comment]
Keep in mind that Node 4 doesn't fully implement ES6, so we're still left with a subset of features we can use. I'd like to propose that we focus on using features that can be checked with our existing rules so we can maximize the dogfoodiness of ESLint.

I think the easiest to start with are `no-var` and `prefer-const` (I'm assuming people want to use `const` everywhere. I'm a bit ambivalent about that, but feel like `prefer-const` is a rule we should be dogfooding.)

**Comment by mysticatea** [comment]
> Keep in mind that Node 4 doesn't fully implement ES6, so we're still left with a subset of features we can use. 

Sure, it's the reason I suggested `node/no-unsupported-features` rule. It would warn it if we use unsupported ES6 features.

It sounds good that we enable those 2 rules to start.

**Comment by platinumazure** [comment]
Wondering if this is worth turning into a milestone, and then we can propose individual features we want to start using as individual issues? (For example, someone could create a rule for "use let/const" which would involve turning on no-var and prefer-const and fixing the resulting lint warnings, and that issue would be added to a "use-es6" milestone.)

**Comment by nzakas** [comment]
I think that's overkill for right now. Let's keep the conversation here until we have some good conclusions and we can figure out next steps after that. 

I'd really like to start using classes, as well.

**Comment by IanVS** [comment]
Does someone with some good familiarity with the capabilities of node@4 want to put together a list of current ESLint rules which we _could_ enable, then we can narrow them down from there?

**Comment by vitorbal** [comment]
@IanVS here you go, a list of all ES6 eslint rules that we *could* enable for node@4, split by feature:

### Arrow functions
- [ ] arrow-body-style: require braces around arrow function bodies
- [ ] arrow-parens: require parentheses around arrow function arguments
- [ ] arrow-spacing: enforce consistent spacing before and after the arrow in arrow functions 
- [ ] no-confusing-arrow: disallow arrow functions where they could be confused with comparisons
- [ ] prefer-arrow-callback: require arrow functions as callbacks

### Generators
- [ ] generator-star-spacing: enforce consistent spacing around * operators in generator functions 
- [ ] require-yield: require generator functions to contain yield
- [ ] yield-star-spacing: require or disallow spacing around the * in yield* expressions 

### Template strings
- [ ] template-curly-spacing: require or disallow spacing around embedded expressions of template strings 
- [ ] prefer-template: require template literals instead of string concatenation

### let/const
- [ ] no-var: require let or const instead of var
- [ ] prefer-const: require const declarations for variables that are never reassigned after declared
- [ ] no-const-assign: disallow reassigning const variables 

### Object literal extensions
- [ ] object-shorthand: require or disallow method and property shorthand syntax for object literals
- [ ] no-useless-computed-key: disallow unnecessary computed property keys in object literals

### Classes
- [ ] constructor-super: require super() calls in constructors 
- [ ] no-class-assign: disallow reassigning class members 
- [ ] no-dupe-class-members: disallow duplicate class members 
- [ ] no-this-before-super: disallow this/super before calling super() in constructors 
- [ ] no-useless-constructor: disallow unnecessary constructors

**Comment by mysticatea** [comment]
@vitorbal I checked some rules which have been enabled by `eslint:recommend` already.

**Comment by platinumazure** [comment]
@mysticatea Note that our repository does not use `eslint:recommended`-- we use `eslint-config-eslint`.

Well, although that should extend `eslint:recommended`? But anyway you might want to take a look at `eslint-config-eslint` to make sure we didn't turn off any recommended rules on this list.

**Comment by nzakas** [comment]
We use `eslint: recommended`: https://github.com/eslint/eslint/blob/master/packages/eslint-config-eslint/default.yml#L2

Thanks @vitorbal for the list. I think we should use all of these. :) What do others think?

**Comment by kaicataldo** [comment]
Is the assumption that we're using the default settings for all of them? If so, using them all is fine by me. I'm :+1: for `arrow-body-style`'s default configuration, but wouldn't be if we were using one of its other configurations.

**Comment by nzakas** [comment]
The stylistic rules need to be discussed, as i don't think just accepting defaults without discussion is a good idea.

**Comment by kaicataldo** [comment]
Ah, okay - thanks, makes sense

**Comment by nzakas** [comment]
I tried converting a single file to use `class` and discovered that our version of PhantomJS doesn't support ES6, which breaks our automated tests. It appears this is the most recent version of the PhantomJS package. :(

**Comment by platinumazure** [comment]
Yes, PhantomJS is hopelessly behind. Are there other test runners we should be considering?

**Comment by mikesherov** [comment]
Why are we using PhantomJS? Unless we're using it to test the browser version (which would require us to transpile *anyway*)?

**Comment by ilyavolodin** [comment]
@mikesherov We have tests for web version of ESLint for website demo. The code is not really transpiled, it's currently just browserified. But if we are starting ES6, we might have to transpile it anyways.

**Comment by mikesherov** [comment]
> But if we are starting ES6, we might have to transpile it anyways.
This is my point exactly. ^

If we are going to test and distribute a browser version, we need to transpile regardless.

**Comment by platinumazure** [comment]
@mikesherov Ah, I think you're saying we can keep using PhantomJS as long as our selected transpiler transpiles to an ES5 that PhantomJS can work with?

**Comment by ilyavolodin** [comment]
Agree. Although I hate transpiling, but not much choice here, really.

**Comment by mysticatea** [comment]
Ah!
I have forgotten about browsers support.
We need to support IE11 also for our online demo. Then, we are using a transpiler "browserify" already. I believe that just adding "babelify" (a transform plugin of browserify) with es2015 preset solves it.

**Comment by nzakas** [comment]
I'd like to hear some other options before going to transpiling. For instance, does anyone know if a PhantomJS version exists with ES6 support? What about the new headless Chrome?

And what do we want to say is our browser support level for the demo?

**Comment by ilyavolodin** [comment]
Some statistics:

Browser Version | Percentage | ES6 support percentage |
|---------------------|---------------|---------------------------------|
|Chrome 50         | 28.29%     | 93% |
|Chrome 49         | 27.93%     | 90% |
|Chrome 51         | 14.73%     | 98% |
|Firefox 45           | 2.62%       | 85% |
|Firefox 46           | 2.13%       | 90% |
|Safari 9               | 5.54%       | 53% |
|IE 11                   | 0.92%       | 15% |

**Comment by hzoo** [comment]
@nzakas Looks like it's in progress https://github.com/ariya/phantomjs/issues/14366 and headless chrome is not ready yet?

**Comment by teppeis** [comment]
PhantomJS v2 (current version) doesn't support ES6 features at all...
See "PJS" column of https://kangax.github.io/compat-table/es6/

**Comment by nzakas** [comment]
I tried throwing babelify into my local setup and it worked pretty easily (literally the first time). So, seems like that may be the way to go.

**Comment by nzakas** [comment]
Can someone with Babel experience take a look at my PR and tell me why the PhantomJS tests are failing? https://github.com/eslint/eslint/pull/6570. I don't have enough energy to keep digging into it.

**Comment by alberto** [comment]
Can we use this issue to discuss about what features we would like to add and where we would use them?

For reference, you can see node@4 ES6 support at http://node.green/

**Comment by nzakas** [comment]
@alberto yup, we can do that. I think we need to include the rules in the discussion because part of the reason to use ES6 features is to dogfood more rules. I think we should err on the side of choices that allow us to enable the most rules.

**Comment by kaicataldo** [comment]
Some easier wins:
* Template literals and the prefer-template rule
* object shorthand notation and the object-shorthand rule (configured with "always")

On a side note,  I'd be willing to go through and make the necessary changes for both of these (most likely in batches to make it easier).

**Comment by nzakas** [comment]
One thing I experimented with is to replace exported objects that act as namespaces with classes that have static methods. Example: https://github.com/eslint/eslint/pull/6570/files#diff-61778371a8bd3b0a5b0ab9c9c92a6cd6L33

I personally like this style because:

1. We can give the class a name that matches how it's used elsewhere.
2. Tests our class parsing and class rules.

However, I'm not married to that idea, so open to hearing what others think.

(Regardless, I do think we should start using classes, even if not for this particular case.)

**Comment by mikesherov** [comment]
I agree that we should use classes but only when we expect to use instances rather than an entire static class, for which I prefer the singleton object pattern we've been using.

Creating a class IMO signifies the intent to use instances. maybe that's just me. Also, if one of the benefits is the ability to name the class, you can do that with objects too:

```
var foo = {};
export default foo;
```

I agree we should be using classes in ESLint, and therefore our rules and parsers will be exercised, but I don't think we should use static only classes for this reason. My 2 cents. I'm not strongly against it though so I wouldn't make another argument if everyone else wants static only classes.

**Comment by mysticatea** [comment]
My 2 cents, I prefer a use of object literal extensions rather than classes in that case.
If it uses classes, I think it's confusing as people may feel that this can be instantiated at the first look.
Object literal extensions allow us enough simple notations.

```js
module.exports = {
    createEmptyConfig() {
        // ....
    },

    // ....
};
```

 I think some class-like functions of [util](https://github.com/eslint/eslint/tree/master/lib/util) would fit for tests our class parsing and class rules.

**Comment by nzakas** [comment]
@kaicataldo why don't you start with the `object-shorthand` rule and we can follow up with the other stuff.

**Comment by vitorbal** [comment]
@kaicataldo May I suggest that we use [this jscodeshift](https://github.com/cpojer/js-codemod/blob/master/transforms/object-shorthand.js) transform to automate that. I've used it before in a pretty large codebase and it saved us a lot of time!

**Comment by kaicataldo** [comment]
@vitorbal Cool! I was planning on using our own autofix - seems like a good use case for it :) Is there a reason why you think we should use the jscodeshift transform instead?

**Comment by nzakas** [comment]
I'd actually suggest we use ESLint autofixing for that to see how far we get.

**Comment by vitorbal** [comment]
You guys are absolutely right, even better if we use our own autofixing for
dogfooding. Don't know why I didn't think of that first :S

**Comment by mgol** [comment]
I don't think rules requiring the use of arrow functions can be enabled in Node 4. While it does support arrows, rest params are available only from Node 5 (now EOL'd) and since arguments is not available in arrow functions that means you can't have arrows with non-specified number of arguments.

**Comment by nzakas** [comment]
@mgol that's just one very narrow use case. I don't think that means we can't use arrow functions.

**Comment by mgol** [comment]
I didn't say you can't use arrow functions but that you can't **require** using them in every case where e.g. the `prefer-arrow-callback` rule would require them. Unless you never use callbacks with indeterminate number of parameters or decide to manually disable ESLint for those lines.

**Comment by alberto** [comment]
Thanks @mgol. That's something to take into account. 👍

**Comment by mysticatea** [comment]
I'm working on this for `no-var` and `prefer-const`.

**Comment by kaicataldo** [comment]
Yikes, this thread is getting really huge. I'll try avoid force pushing so much to try to keep this issue easier to manage. Wanted to follow up and see what everyone else though about enabling the final rule in [my comment](https://github.com/eslint/eslint/issues/6407#issuecomment-231392810) above, which is `prefer-template`. There isn't an autofix for this, so some kind of codemod might be what we want here...

**Comment by nzakas** [comment]
I'm 👍. More rules, more fun. :)
